File: binary-tree-zigzag-level-order-traversal/binary-tree-zigzag-level-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        Queue<TreeNode> queue = new LinkedList<>();
        List<List<Integer>> result = new ArrayList<>();
        if (root == null)
            return result;
        queue.add(root);
        
        while (!queue.isEmpty()) {
            int size = queue.size();
            boolean leftToRight = result.size() % 2 == 0;
            LinkedList<Integer> level = new LinkedList<>();
            for (int i = 0; i < size; i++) {
                TreeNode node = queue.poll();
                if (leftToRight) 
                    level.add(node.val);
                else
                    level.addFirst(node.val);
                if (node.left != null)
                    queue.add(node.left);
                if (node.right != null)
                    queue.add(node.right); 
            }
            result.add(level);
        }
        return result;
    }
}

File: binary-tree-zigzag-level-order-traversal/binary-tree-zigzag-level-order-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from collections import deque
class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        queue = deque()
        result = []
        if root is None:
            return result
        queue.append(root)
        i = 0
        while queue:
            size = len(queue)
            level = []
            for _ in range(size):
                node = queue.popleft()
                if node.left is not None:
                    queue.append(node.left)
                if node.right is not None:
                    queue.append(node.right)
                level.append(node.val)
            if i % 2 == 1:
                level.reverse()
            i += 1
            result.append(level)
        return result
            
        